cep-0021.md

```diff
@@ -0,0 +1,229 @@
+
+The channel and subdir MUST be used as-is without modification.
+
+The channel, subdir and OCI-encoded package name (or hashed OCI-encoded package name if required) MUST be combined with `/` to form the OCI repository `<name>`
+
+```
+<channel>/<subdir>/<OCI-encoded package name>
+```
+
+#### Encoding Labels, Versions and Build Strings into OCI `<tag>`s
+
+The label, version, and build string are encoded to OCI-form as follows. First the following mapping rules MUST be applied to the label, version and build string in the order listed below:
+
+- `_` -> `_U`
+- `-` -> `_D`
+- `+` -> `_P`
+- `!` -> `_N`
+- `=` -> `_E`
+- `:` -> `_C`
+- `/` -> `_S`
+- ` ` -> `_B`
+- `\t` -> `_T`
+- `\r` -> `_R`
+- `\n` -> `_L`
+
+This encoding can be undone by applying the rules in reverse, starting at the bottom of the list and moving to the top. Depending on the context, some labels may be percent-encoded for use in URLs. The percent-encoding MUST be undone before the label is encoded via the list of rules above.
+
+After encoding each component to OCI-form, the version, build string, and (optional) label MUST be combined with a `-` to form the OCI `<tag>` as either
+
+```
+<version>-<build>-<label>
+```
+
+or if the label is not specified (i.e. the package is on the `main` label), then
+
+```
+<version>-<build>
+```
+
+with no trailing `-`.
+
+As implied above, if no label is explicitly specified for a package, then package is by definition on the `main` label. This stipulation means the following two OCI tags for conda packages are equivalent:
+
+```
+v1.0.0-h123456__0-main
+v1.0.0-h123456__0
+```
+
+Finally, if the entire OCI tag exceeds 128 characters in length, the entire tag MUST be hashed with the SHA1 algorithm and the resulting OCI tag is then `h<SHA1 in hexadecimal>`.
+
+#### Further Odds and Ends
+
+- If either the OCI-encoded form of package name or the OCI tag for the conda package are hashed, then both MUST be hashed. In other words, only the two forms of the OCI artifact `<name>:<tag>` combination defined above, either unhashed or hashed, are allowed.
+- Once the OCI repository `<name>` and `<tag>` are computed for the conda artifact, the rest of the OCI Distribution Spec MUST be followed.
+- If an OCI registry implementation errors due to length when dealing with the full OCI form of the conda package (e.g., the combination of the registry URL and full OCI repository `<name>` for the conda package exceed some internal limit in the implementation), then that package cannot be stored in that registry and an error MUST be raised.
+- Within conda repodata, the download URL for a conda package in an OCI registry MUST always be prepended by `oci://` to indicate that the download URL is an OCI URL. The registry being used, any required ports etc. are prepended to the conda package OCI repository `<name>:<tag>` to form the full download URL per the OCI v1.0 specification.
```

Comment:
```suggestion
- The download URL for a conda package in an OCI registry MUST use the `oci://` scheme to indicate that the artifact will be downloaded from an OCI endpoint. More specifically, it MUST follow the syntax `oci://<authority>[:<port>]/[<path>/]<name>:<tag>`.
```

Reply:
There would be no extra <path> component here as far as I know, right?

Reply:
Nope, just what you suggested but using the established URL terms (authority, path, etc).